README.md
# chip-io

![logo](images/CHIP-J5.png)

[Johnny-Five](https://github.com/rwaldron/johnny-five) [IO Plugin](https://github.com/rwaldron/io-plugins) for the [Next Thing Co.](http://nextthing.co/index.html) [C.H.I.P.](http://getchip.com)

## Prerequisites

 * Next Thing Co. [C.H.I.P.](http://getchip.com) board
 * Node.js installed
   1. Install ```curl```: ```sudo apt-get install curl```
   2. Follow Debian section of [NodeSource installations instructions](https://github.com/nodesource/distributions#debinstall)
 * Build essential installed: ```sudo apt-get install build-essential```
 * Add ```chip``` user to ```i2c``` group: ```sudo adduser chip i2c```

__NOTE:__ Some GPIO pins require `root` or `sudo`.

## Getting Started

```sh
npm install chip-io johnny-five
```

__NOTE:__ Must be run on the C.H.I.P. itself.

## Boilerplate Program

```javascript
var five = require('johnny-five');
var chipio = require('chip-io');

var board = new five.Board({
  io: new chipio()
});

board.on('ready', function() {
  // do Johnny-Five stuff
});
```

## Examples

See [examples](examples) folder as well as [Johnny-Five examples](http://johnny-five.io/examples/).

## API

See  [Johnny-Five API docs](http://johnny-five.io/api/).

## Pin Guide

| Johnny-Five Compatible Name | Number | Supported Modes | Requires root or sudo | Info |
| --------------------------- | :----: | --------------- | :-------------------: | ---- |
| LCD-D2 | 17 | Input | ✓ |  |
| PWM0 | 18 | Input, Output, PWM | ✓ |  |
| LCD-D4 | 19 | Input, Output | ✓ |  |
| LCD-D3 | 20 | Input, Output | ✓ |  |
| LCD-D6 | 21 | Input, Output | ✓ |  |
| LCD-D5 | 22 | Input, Output | ✓ |  |
| LCD-D10 | 23 | Input, Output | ✓ |  |
| LCD-D7 | 24 | Input, Output | ✓ |  |
| LCD-D12 | 25 | Input, Output | ✓ |  |
| LCD-D11 | 26 | Input, Output | ✓ |  |
| LCD-D14 | 27 | Input, Output | ✓ |  |
| LCD-D13 | 28 | Input, Output | ✓ |  |
| LCD-D18 | 29 | Input, Output | ✓ |  |
| LCD-D15 | 30 | Input, Output | ✓ |  |
| LCD-D20 | 31 | Input, Output | ✓ |  |
| LCD-D19 | 32 | Input, Output | ✓ |  |
| LCD-D22 | 33 | Input, Output | ✓ |  |
| LCD-D21 | 34 | Input, Output | ✓ |  |
| LCD-CLK | 35 | Input, Output | ✓ |  |
| LCD-D23 | 36 | Input, Output | ✓ |  |
| LCD-VSYNC | 37 | Input, Output | ✓ |  |
| LCD-HSYNC | 38 | Input, Output | ✓ |  |
| LCD-DE | 40 | Input, Output | ✓ |  |
| LRADC | 51 | Analog | ✓ |  |
| XIO-P0 | 53 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P1 | 54 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P2 | 55 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P3 | 56 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P4 | 57 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P5 | 58 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P6 | 59 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| XIO-P7 | 60 | Input, Output |  | Connected to the built-in PCF8574A IO extender |
| CSIPCK | 67 | Input | ✓ |  |
| CSICK | 68 | Input | ✓ |  |
| CSIHSYNC | 69 | Input | ✓ |  |
| CSIVSYNC | 70 | Input, Output | ✓ |  |
| CSID0 | 71 | Input, Output | ✓ |  |
| CSID1 | 72 | Input, Output | ✓ |  |
| CSID2 | 73 | Input, Output | ✓ |  |
| CSID3 | 74 | Input, Output | ✓ |  |
| CSID4 | 75 | Input, Output | ✓ |  |
| CSID5 | 76 | Input, Output | ✓ |  |
| CSID6 | 77 | Input, Output | ✓ |  |
| CSID7 | 78 | Input, Output | ✓ |  |
| I2C | | I2C | | Uses I2C port 2 (TWI2-SCK and TWI2-SDA). Address 0x38 is used by the built-in PCF8574A IO extender |

![C.H.I.P. pinouts](http://docs.getchip.com/images/chip_pinouts.jpg)

## Additional Features

| Type | Usage | Johnny-Five type | Notes |
| ---- | ----- | ---------------- | ----- |
| Battery Voltage | `new chipio.BatteryVoltage();` | [five.Sensor](http://johnny-five.io/api/sensor/) | Reads battery voltage from the AXP290 |
| Internal Temperature | `new chipio.InternalTemperature();` | [five.Thermometer](http://johnny-five.io/api/thermometer/) | Reads internal temperature from the AXP290 |
| Onboard Button | `new chipio.OnboardButton();` | [five.Button](http://johnny-five.io/api/button/) | Use onboard button (beside `LCD-DE` pin) connected to the AXP290 |
| Status LED | `new chipio.StatusLed;` | [five.Led](http://johnny-five.io/api/led/) | Controls status LED connected to GPIO2 on the AXP290 |


CHANGELOG.md
## Version 2.1.1

 * Added new on-board button and status LED example. ([@norgeous](https://github.com/norgeous))
 * Switch to `i2c-bus` module for I2C layer. ([@fivdi](https://github.com/fivdi))
 * Throw exception for `pingRead`.

## Version 2.1.0

 * Fixed I2C address 0 not working
 * Added support for Onboard Button

## Version 2.0.2

 * Correct internal temperature controller name.
 * Correct PWM polarity inversion.
 * Switch to sync. I/O, resolves I2C issues.

## Version 2.0.1

 * Correct typeof comparison for addressToBus map, fixes I2C read and write operations.

## Version 2.0.0

 * Officially added support to access LCD pins via `mmap` to `/dev/mem` (requires `root` or `sudo`).
 * Added PWM and LRADC support.
 * Added new components for battery voltage, internal temperature, and status LED.
 * Fixed segfault if `sudo` was not used for R8 pins. ([@urish](https://github.com/urish))
 * Bus 2 is now the default bus for I2C. ([@bbx10](https://github.com/bbx10))
 * Added support for J2 components to specificy I2C bus via `i2cConfig` ([@rwaldron](https://github.com/rwaldron))

## Older

 * Changes not recorded


lib/chip-io.js
var util = require('util');

var BoardIO = require('board-io');
var debug = require('debug')('chip-io');
var i2cBus = require('i2c-bus');

var AllwinnnerR8 = require('./allwinner-r8');
var AXP209 = require('./axp209');
var PCF8574A = require('./pcf8574a');

var pins = require('./pins');

var TICK_INTERVAL = 19;
var DEFAULT_BUS = 2;
var addressToBus = {};

function ChipIO() {
  // call super constructor
  BoardIO.call(this, {
    quiet: true
  });

  this.name = 'C.H.I.P.';
  this.defaultLed = 'STATUS';

  this._pinsById = {};

  var analogChannel = 0;

  pins.forEach(function(pin, index) {
    pin.number = index;

    if (pin.supportedModes.length > 0) {
      pin.id = pin.name;

      if (pin.supportedModes.indexOf(2) !== -1) {
        pin.analogChannel = analogChannel;

        analogChannel++;
      }

      this._pinsById[pin.id] = pin;
    }

    this._pins.push(pin);
  }.bind(this));

  // emit 'connect' event on the next tick
  process.nextTick(function() {
    this.emit('connect');
  }.bind(this));

  this._i2cBuses = {};

  this._chips = {
    AXP209: new AXP209(0, 0x34),
    PCF8574A: new PCF8574A(2, 0x38),
    R8: new AllwinnnerR8()
  };

  this._chips.AXP209.on('analog-read', this._onAnalogRead.bind(this));
  this._chips.AXP209.on('digital-read', this._onDigitalRead.bind(this));
  this._chips.PCF8574A.on('digital-read', this._onDigitalRead.bind(this));
  this._chips.R8.on('analog-read', this._onAnalogRead.bind(this));
  this._chips.R8.on('digital-read', this._onDigitalRead.bind(this));

  this._chips.AXP209.open();
  this._chips.PCF8574A.open();

  try {
    this._chips.R8.open();
  } catch(err) {
    // "optional" (requires root), so ignore err
  }

  // all done, emit 'ready' event on the next tick
  process.nextTick(function() {
    this.emit('ready');

    // start the internal ticker
    this._tick();
  }.bind(this));
}

util.inherits(ChipIO, BoardIO);

ChipIO.prototype.normalize = function(pin) {
  if (typeof(pin) === 'string') {
    pin = this._pinsById[pin].number;
  }

  return pin;
};

ChipIO.prototype.pinMode = function(pin, mode) {
  debug('pinMode', pin, mode);

  pin = this.normalize(pin);

  var pinData = this._pins[pin];

  if (pinData && pinData.chip) {
    this._pins[pin].mode = mode;

    this._chips[pinData.chip].pinMode(pinData.name, mode);
  }

  return this;
};

ChipIO.prototype.pwmWrite = function(pin, value) {
  debug('pwmWrite', pin, value);

  pin = this.normalize(pin);

  var pinData = this._pins[pin];

  if (pinData && pinData.chip) {
    this._chips[pinData.chip].pwmWrite(pinData.name, value);
  }

  return this;
};

ChipIO.prototype.analogWrite = ChipIO.prototype.pwmWrite;

ChipIO.prototype.analogRead = function(pin, handler) {
  debug('analogRead', pin);

  pin = this.normalize(pin);

  var pinData = this._pins[pin];

  if (pinData && pinData.chip) {
    this._pins[pin].report = 1;

    this.on('analog-read-' + pin, handler);

    this._chips[pinData.chip].analogRead(pinData.name);
  }

  return this;
};

ChipIO.prototype.digitalWrite = function(pin, value) {
  debug('digitalWrite', pin, value);

  pin = this.normalize(pin);

  var pinData = this._pins[pin];

  if (pinData && pinData.chip) {
    this._chips[pinData.chip].digitalWrite(pinData.name, value);
  }

  return this;
};

ChipIO.prototype.digitalRead = function(pin, handler) {
  debug('digitalRead', pin);

  pin = this.normalize(pin);

  var pinData = this._pins[pin];

  if (pinData && pinData.chip) {
    this._pins[pin].report = 1;

    this.on('digital-read-' + pin, handler);

    this._chips[pinData.chip].digitalRead(pinData.name);
  }

  return this;
};

ChipIO.prototype.i2cConfig = function(options) {
  debug('i2cConfig', options);

  // note that there's a design flaw here
  // two devices with the same address on different buses doesn't work
  // see https://github.com/rwaldron/io-plugins/issues/13

  // options.address is _always_ sent by all I2C component classes in
  // Johnny-Five
  var address = options.address;

  // options.bus is optional
  var bus = typeof(options.bus) !== 'undefined' ? options.bus : DEFAULT_BUS;

  // associate the address to the bus
  if (!addressToBus.hasOwnProperty(address)) {
    addressToBus[address] = bus;
  }

  // create an i2c-bus object for the I2C bus
  if (!this._i2cBuses.hasOwnProperty(bus)) {
    this._i2cBuses[bus] = i2cBus.openSync(bus);
  }

  return this;
};

ChipIO.prototype.i2cWrite = function(address, cmdRegOrData, inBytes) {
  debug('i2cWrite', address, cmdRegOrData, inBytes);

  var i2c = this._i2cBuses[addressToBus[address]];

  // if i2cWrite was used for an i2cWriteReg call...
  if (arguments.length === 3 &&
      !Array.isArray(cmdRegOrData) &&
      !Array.isArray(inBytes)) {
    return this.i2cWriteReg(address, cmdRegOrData, inBytes);
  }

  // fix arguments if called with Firmata.js API
  if (arguments.length === 2) {
    if (Array.isArray(cmdRegOrData)) {
      inBytes = cmdRegOrData.slice();
      cmdRegOrData = inBytes.shift();
    } else {
      inBytes = [];
    }
  }

  var buffer = new Buffer([cmdRegOrData].concat(inBytes));

  // only write if bytes provided
  if (buffer.length) {
    i2c.i2cWriteSync(address, buffer.length, buffer);
  }
  return this;
};

ChipIO.prototype.i2cWriteReg = function(address, register, byte) {
  debug('i2cWriteReg', address, register, byte);

  var i2c = this._i2cBuses[addressToBus[address]];

  i2c.writeByteSync(address, register, byte);

  return this;
};

ChipIO.prototype.i2cRead = function(address, register, size, handler) {
  debug('i2cRead', address, register, size, handler);

  var continuousRead = function() {
    this.i2cReadOnce(address, register, size, function(bytes) {
      handler(bytes);
      setTimeout(continuousRead, TICK_INTERVAL);
    });
  }.bind(this);

  continuousRead();

  return this;
};

ChipIO.prototype.i2cReadOnce = function(address, register, size, handler) {
  // fix arguments if called with Firmata.js API
  if (arguments.length === 3 &&
      typeof register === "number" &&
      typeof size === "function") {
    handler = size;
    size = register;
    register = null;
  }

  debug('i2cReadOnce', address, register, size, handler);

  var event = "I2C-reply" + address + "-" + (register !== null ? register : 0);

  var afterRead = function (err, bytesRead, buffer) {
    if (err) {
      return this.emit("error", err);
    }

    // convert buffer to an Array before emit
    this.emit(event, Array.prototype.slice.call(buffer));
  }.bind(this);

  if (typeof handler === "function") {
    this.once(event, handler);
  }

  var i2c = this._i2cBuses[addressToBus[address]];
  var data = new Buffer(size);

  if (register !== null) {
    i2c.readI2cBlock(address, register, size, data, afterRead);
  } else {
    i2c.i2cRead(address, size, data, afterRead);
  }

  return this;
};

ChipIO.prototype.pingRead = function() {
  throw new Error('pingRead is not supported! Please use an I2C backpack: http://johnny-five.io/examples/proximity-hcsr04-i2c/');
};

ChipIO.prototype._tick = function() {
  var asyncTicksRunning = 2;

  var afterTick = function() {
    asyncTicksRunning -= 1;

    if (asyncTicksRunning === 0) {
      // schedule next tick
      this._tickTimeout = setTimeout(this._tick.bind(this), TICK_INTERVAL);
    }
  }.bind(this);

  this._chips.AXP209.tick(afterTick);
  this._chips.PCF8574A.tick(afterTick);
  this._chips.R8.tick();
};

ChipIO.prototype._onAnalogRead = function(id, value) {
  var pin = this._pinsById[id];

  this.emit('analog-read-' + pin.number, value);
};

ChipIO.prototype._onDigitalRead = function(id, value) {
  var pin = this._pinsById[id];

  this.emit('digital-read-' + pin.number, value);
};

module.exports = ChipIO;
